stage #2/module #1 Modern Java/lesson #3. Old services reborn/content.md

```diff
@@ -0,0 +1,750 @@
+    
+    HttpResponse<String> response = HttpClient.newHttpClient()
+    .send(request, HttpResponse.BodyHandler.asString());
+    CompletableFuture<HttpHeaders> trailers = response.trailers();
+
+<b>Java 11 Http Client</b>
+
+The major change in Java 11 was the standardization of HTTP client API that implements HTTP/2 and Web Socket. It aims to replace the legacy HttpUrlConnection class which has been present in the JDK since the very early years of Java.
+
+The main changes:
+
++ The newer version of the HTTP protocol is designed to improve the overall performance of sending requests by a client and receiving responses from the server. This is achieved by introducing a number of changes such as stream multiplexing, header compression and push promises.
++ As of Java 11, <b>*the API is now fully asynchronous (the previous HTTP/1.1 implementation was blocking)*</b>. Asynchronous calls are implemented using  CompletableFuture. The CompletableFuture implementation takes care of applying each stage once the previous one has finished, so this whole flow is asynchronous.
++ The new APIs provide native support for HTTP 1.1/2 WebSocket. The core classes and interface providing the core functionality include:
+  + java.net.http.HttpClient
+  + java.net.http.HttpRequest
+  + java.net.http.HttpResponse
+  + java.net.http.WebSocket
+
+*Introduction of Static Factory Classes*
+
+New static factory classes BodyPublishers, BodySubscribers, and BodyHandlers are introduced that include existing implementations of BodyPublisher, BodySubscriber and BodyHandler.
+
+These are used to perform useful common tasks, such as handling the response body as a String or streaming the body to a File.
+
+For e.g. in Pre Java 11 we had to do something like this:
+
+    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandler.asString());
+
+Which we can now simplify as:
+
+    HttpResponse<String> response = client.send(request, BodyHandlers.ofString());
+
+*Handling Push Promises in HTTP/2*
+
+New Http client supports push promises through PushPromiseHandler interface.
+
+<b>*It allows the server to “push” content to the client additional resources while requesting the primary resource, saving more roundtrip and as a result, improves performance in page rendering.*</b>
+
+It is really the multiplexing feature of HTTP/2 that allows us to forget about resource bundling. For each resource, the server sends a special request, known as a push promise to the client.
+
+Push promises received, if any, are handled by the given PushPromiseHandler. A null valued PushPromiseHandler rejects any push promises.
+
+The HttpClient has an overloaded sendAsync method that allows us to handle such promises, as shown in the below example.
+
+Let's first create a PushPromiseHandler:
+
+    private static PushPromiseHandler<String> pushPromiseHandler() {
```

Comment:
good method for asking on the interview 'what is going on here?' Just don't know for which level..

Reply:
discussed with AliakseiYermachonak, decided to leave as is